Source/WebKit/UIProcess/WebPageProxy.cpp

```diff
@@ -6342,6 +6343,9 @@ void WebPageProxy::decidePolicyForNavigationAction(Ref<WebProcessProxy>&& proces
         }
         if (!navigation)
             navigation = m_navigationState->createLoadRequestNavigation(process->coreProcessIdentifier(), ResourceRequest(request), m_backForwardList->currentItem());
+
+        if (fromAPI && !navigationID && !internals().pageLoadState.pendingAPIRequestURL().isNull())
```

Comment:
Why can't we do the following instead?

In `WebPageProxy::receivedPolicyDecision()`, if the policyDecision is `Ignore`, clear the pending API request URL (after checking that the policyDecision was for the same navigation as the pending API request).

Reply:
Are you suggesting that we should compare URLs instead of navigationIDs? This sounds very similar to the current behavior.

Reply:
No I didn’t suggest comparing URLs. I suggested checking navigation ids but in a different place and only when the policy decision is “ignore”.

Reply:
I'm sorry, that sounds like what we already have in that function, am I misunderstanding and you are proposing something else?
```
    if (action == PolicyAction::Ignore && willContinueLoadInNewProcess == WillContinueLoadInNewProcess::No && navigation && navigation->navigationID() == internals().pageLoadState.pendingAPIRequest().navigationID)                                
        internals().pageLoadState.clearPendingAPIRequest(transaction);
```

Reply:
Ah, we do have it indeed. Which brings the question, why isn't your API test not passing with the existing code? Why are we not clearing the pending API request even though the client said to ignore the navigation?

Reply:
That test isn't passing because when we restore state, we don't create an API::Navigation when we load the "current item" from the saved state. In that case, we set the pending API request's navigationID as 0:
```
    RefPtr<API::Navigation> navigation;
    if (!m_backForwardList->currentItem()->itemIsInSameDocument(item))
        navigation = m_navigationState->createBackForwardNavigation(process().coreProcessIdentifier(), item, m_backForwardList->currentItem(), frameLoadType);

    auto transaction = internals().pageLoadState.transaction();
    internals().pageLoadState.setPendingAPIRequest(transaction, { navigation ? navigation->navigationID() : 0, item.url() });
```
Then, later, when we decide the navigation policy, we create a new API::Navigation. Therefore the cached pending API request's navigationID and the new Navigation's ID won't match and we won't clear the pending API request. Alternatively, we could change the behavior in `WebPageProxy::goToBackForwardItem` instead of fixing-up the navigationID (like this patch proposes).

Reply:
Seems we should consider updating the session restoration code to create a Navigation object if possible? I'd rather we address the problem at the root.

Reply:
Sure, I'll do that. This patch seemed safer in case there are other similar cases where we don't create a Navigation, but I'll make a more targeted fix.